packages/flutter_driver/lib/src/driver/web_driver.dart

```diff
@@ -0,0 +1,270 @@
+    //       UI    : <-- build -->
+    //       GPU   :               <-- rasterize randomly slow today -->
+    //       Gap   :              |    2 seconds or more   |
+    //       Driver:                                        <-- screenshot -->
+    //
+    //       In practice, sometimes the device gets really busy for a while and
+    //       even two seconds isn't enough, which means that this is still racy
+    //       and a source of flakes.
+    await Future<void>.delayed(const Duration(seconds: 2));
+
+    return _connection.screenshot();
+  }
+
+  @override
+  Future<void> startTracing({
+    List<TimelineStream> streams = const <TimelineStream>[TimelineStream.all],
+    Duration timeout = kUnusuallyLongTimeout,
+  }) async {
+    if (_browserName != kChrome) {
+      throw UnimplementedError();
+    }
+
+    _startTime = DateTime.now();
+  }
+
+  @override
+  Future<Timeline> stopTracingAndDownloadTimeline({Duration timeout = kUnusuallyLongTimeout}) async {
+    if (_browserName != kChrome) {
+      throw UnimplementedError();
+    }
+    if (_startTime == null) {
+      return null;
+    }
+
+    final List<Map<String, dynamic>> events = <Map<String, dynamic>>[];
+    for (sync.LogEntry entry in _connection.logs) {
+      if (_startTime.isBefore(entry.timestamp)) {
+        final Map<String, dynamic> data = jsonDecode(entry.message)['message'];
+        if (data['method'] == 'Tracing.dataCollected') {
+          // 'ts' data collected from Chrome is in double format, conversion needed
+          data['params']['ts'] =
+              double.parse(data['params']['ts'].toString()).toInt();
+          events.add(data['params']);
+        }
+      }
+    }
+    final Map<String, dynamic> json = <String, dynamic>{
+      'traceEvents': events,
+    };
+    _startTime = null;
+    return Timeline.fromJson(json);
+  }
+
+  @override
+  Future<Timeline> traceAction(Future<dynamic> Function() action, {
+    List<TimelineStream> streams = const <TimelineStream>[TimelineStream.all],
+    bool retainPriorEvents = false,
+  }) async {
+    if (_browserName != kChrome) {
+      throw UnimplementedError();
+    }
+    if (!retainPriorEvents) {
+      await clearTimeline();
+    }
+    await startTracing(streams: streams);
+    await action();
+
+    return stopTracingAndDownloadTimeline();
+  }
+
+  @override
+  Future<void> clearTimeline({Duration timeout = kUnusuallyLongTimeout}) async {
+    if (_browserName != kChrome) {
+      throw UnimplementedError();
+    }
+
+    // Reset start time
+    _startTime = null;
+  }
+}
+
+/// Encapsulates connection information to an instance of a Flutter Web application.
+class FlutterWebConnection {
+  FlutterWebConnection._(this._driver);
+
+  final sync.WebDriver _driver;
+
+  /// Starts WebDriver with the given [capabilities] and
+  /// establishes the connection to Flutter Web application.
+  static Future<FlutterWebConnection> connect(String url, Map<String, dynamic> settings) async {
```

Comment:
This should perhaps take a configurable timeout parameter, defaulting to no timeout